simple_certmanager/forms.py

```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
             instance.save()
         return instance
 
+    def clean(self) -> dict[str, Any]:
+        # Errors if the CSR already has a certificate
+        if self.instance.public_certificate:
+            self.add_error(
+                "certificate",
+                "A certificate already exists for this CSR. "
+                "Delete the certificate first.",
+            )
+        # Handle the certificate file upload
+        if "certificate" in self.cleaned_data and self.cleaned_data["certificate"]:
+            uploaded_file = self.cleaned_data["certificate"]
+            try:
+                uploaded_file.seek(0)
+                certificate_data = uploaded_file.read()
+
+                # Load the certificate and throw an error if the certificate is invalid
+                certificate = x509.load_pem_x509_certificate(certificate_data)
+                private_key = load_pem_x509_private_key(
+                    self.instance.private_key.encode()
+                )
+
+                # Check if the certificate matches the CSR by comparing the public keys
+                private_key_pubkey = private_key.public_key()
+                certificate_public_key = certificate.public_key()
+                match = private_key_pubkey == certificate_public_key
+                if not match:
+                    self.add_error(
+                        "certificate",
+                        "Certificate does not match the signature from the actual CSR.",
+                    )
+
+            except ValueError:
+                self.add_error(
+                    "certificate", "Invalid certificate. Check the file format."
+                )
+            except Exception:
+                self.add_error("certificate", "An unexpected error occurred.")
```

Comment:
```suggestion
```
it's a bad idea to display programming errors to the user as if they're form validation errors.

If an unexpected error happens, then we need to see that it and what happened in Sentry, and suppressing exceptions is usually a bad idea for that reason.